srml/rolling-window/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Rolling Window module
+//!
+//! ## Overview
+//!
+//! The Rolling Window Module is similar to `simple moving average` except
+//! that it just reports the number of occurrences in the window instead of
+//! calculating the average.
+//!
+//! It is mainly implemented to keep track of misbehaviors and only the take
+//! the last `sessions` of misbehaviors into account.
+//!
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(test)]
+mod mock;
+
+use srml_support::{
+	StorageValue, StorageMap, EnumerableStorageMap, decl_module, decl_storage,
+	traits::WindowLength
+};
+use parity_codec::Codec;
+use sr_primitives::traits::MaybeSerializeDebug;
+
+type Session = u32;
+
+/// Rolling Window trait
+pub trait Trait: system::Trait {
+	/// Kind to report with window length
+	type Kind: Copy + Clone + Codec + MaybeSerializeDebug + WindowLength<u32>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as RollingWindow {
+		/// Misbehavior reports
+		///
+		/// It maps each kind into a hash (identity of reporter) and session number when it occurred
+		MisconductReports get(kind): linked_map T::Kind => Vec<(Session, T::Hash)>;
+
+		/// Session index
+		SessionIndex get(s) config(): u32 = 0;
+	}
+}
+
+decl_module! {
+	/// Rolling Window module
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {}
+}
+
+impl<T: Trait> Module<T> {
+	/// Remove items that doesn't fit into the rolling window
+	/// Must be called when a session ends
+	pub fn on_session_end() {
+		let session = SessionIndex::get().wrapping_add(1);
+
+		for (kind, _) in <MisconductReports<T>>::enumerate() {
+			let window_length = kind.window_length();
+			<MisconductReports<T>>::mutate(kind, |reports| {
+				// it is guaranteed that `reported_session` happened before `session`
+				reports.retain(|(reported_session, _)| {
+					let diff = session.wrapping_sub(*reported_session);
+					diff < *window_length
+				});
+			});
+		}
+
+		SessionIndex::put(session);
+	}
+
+	/// Report misbehavior for a kind
+	pub fn report_misbehavior(kind: T::Kind, footprint: T::Hash) {
+		let session = SessionIndex::get();
+
+		if <MisconductReports<T>>::exists(kind) {
+			<MisconductReports<T>>::mutate(kind, |w| w.push((session, footprint)));
+		} else {
+			<MisconductReports<T>>::insert(kind, vec![(session, footprint)]);
+		}
+	}
+
+	/// Return number of misbehavior's in the current window which
+	/// may include duplicated misbehaviour's
+	pub fn get_misbehaved(kind: T::Kind) -> u64 {
+		<MisconductReports<T>>::get(kind).len() as u64
+	}
+
+
+	/// Return number of misbehavior's in the current window which
+	/// ignores duplicated misbehavior's in each session
+	pub fn get_misbehaved_unique(kind: T::Kind) -> u64 {
```

Comment:
this function is missing the mark on what we want from uniqueness.

1. uniqueness should be detected on insertion, not on query. attempting to insert a non-unique misbehavior should lead to failure (which is why `report_misbehavior` should return a `Result`).
2. uniqueness should be enforced for the entire bonding period, not just a rolling window.

----

With that in mind, I think enforcing uniqueness should be done like this:
1. Create a new entry in storage mapping `Hash => Session`
2. When reporting, insert if and only if there is no entry there. The session is the current session index.
3. We don't prune for the moment, so it will grow unbounded. but storing the session index will let us do pruning easily later (we will have to think of how to reward it -- probably by re-minting coins to someone who issues a transaction to prune it when it's old)

Reply:
> uniqueness should be enforced for the entire bonding period,

Well, just to verify that I understand this correctly

bonding  period == bond(stash, controller) ------> unbond(controller) -------> BondingDuration

I'm not sure how to keep track of the `bonding period`. Is it sufficient to just have a event handler `on_bonding_period_end()` to be called and maybe reset it accordingly?

> We don't prune for the moment, so it will grow unbounded. but storing the session index will let us do pruning easily later (we will have to think of how to reward it -- probably by re-minting coins to someone who issues a transaction to prune it when it's old)

I don't understand what we are not `pruning` can you explain?

Reply:
See the second list of 3 items. Just implement that

Reply:
@niklasad1 i think @rphmeier means that for now, any given account id must never report the same misbehaviour more than once, regardless of how far apart they are. so you should just store the hash of the report as the key of a storage item (session number at which it was reported would be a sensible value) and quickly look it up to make sure it's empty when a misbehaviour is reported. if it's not empty then it should return an `Err`. 

at some later stage we can create logic that knows about how long the bonding period is in order to allow these storage items to be removed.

Reply:
thanks Gav, @rphmeier clarified it for me yesterday.

I have implemented logic to remove `misbehaviors` older than one bonding period where the `BondingDuration` is an associated type (not really coupled to staking). 

So, if Staking::BondingDuration != Self::BondingDuration it could we could get unexpected behavior. 

I guess we could require `RollingWindow` to inherit the `Staking trait` otherwise I don't think it is possible to fetch but I would like to avoid that.

Thoughts?